hphp/runtime/vm/jit/relocation-arm.cpp

```diff
@@ -0,0 +1,993 @@
+#include "hphp/runtime/vm/jit/asm-info.h"
+#include "hphp/runtime/vm/jit/cg-meta.h"
+#include "hphp/runtime/vm/jit/containers.h"
+#include "hphp/runtime/vm/jit/fixup.h"
+#include "hphp/runtime/vm/jit/ir-opcode.h"
+#include "hphp/runtime/vm/jit/smashable-instr.h"
+
+#include "hphp/vixl/a64/macro-assembler-a64.h"
+
+namespace HPHP { namespace jit { namespace arm {
+
+using namespace vixl;
+
+namespace {
+
+TRACE_SET_MOD(mcg);
+
+//////////////////////////////////////////////////////////////////////
+
+/*
+ * Support for relocation on ARM involves implementing the following:
+ *
+ *   adjustForRelocation - Adjusts a range of instructions which may
+ *                         contain addresses which have been moved.
+ *   adjustCodeForRelocation - Similar to adjustForRelocation, but is
+ *                             reserved for ranges which may still
+ *                             hold live instructions.
+ *   adjustMetaDataForRelocation - Adjusts the affected metadata
+ *                                 following a relocation.
+ *   findFixups - Finds the affected fixups following a relocation.
+ *   relocate - Relocates a range of instructions to a new
+ *              destination. Critically, relocation is a chance
+ *              to grow/shrink the region which is advantageous
+ *              when moving a far branch to simple near relative
+ *              branch for instance.
+ *
+ * adjustForRelocation and adjustCodeForRelocation both search for
+ * addresses in instruction ranges that need to be updated following
+ * a relocation. This involves finding instructions which might
+ * be loading an address and then determining if that target is
+ * actually an address for an instruction which has been moved. For
+ * ARM, this includes the following:
+ *
+ *   PC Relative
+ *     ADR/ADRP - Builds PC relatative addresses.
+ *     B[.<cc>] (immediate) - Branch to a PC relative address.
+ *     LDR (literal) - Loads a literal from a PC relative address.
+ *
+ *   Immediates
+ *     MOV/MOK - Used to build up targets 16 bits at a time.
+ *     LDR (literal) - The literal loaded may be a target address.
+ *
+ * adjustCodeForRelocation must support updating live instructions.
+ * For ARM, we can update a single instruction or a 64 bit literal
+ * in the instruction stream atomically. We can't, however, update
+ * multiple instructions atomically since there may be a thread
+ * executing instructions within that range. Therefore, MOV/MOVK
+ * instructions can't be adjusted for live ranges of instructions.
+ * Any emitted code that uses MOV/MOVK and which must be adjusted
+ * while live, should instead use LDR (literal). Smashables use this
+ * approach; they embed a literal in the instruction stream which
+ * can then be updated atomically. At the time of this writing, only
+ * the vasm opcode (leap) produces a live region that must be updated,
+ * and so it uses an approach with LDR (literal) instead of MOV/MOVK.
```

Comment:
Thanks for documenting all of this!  A lot of it is well-suited to accompany the APIs that are laid out in `relocation.h`.  Can you move the non-ARM-specific bits there, and leave the rest here?